core/scratch_block_comment.js

```diff
@@ -0,0 +1,404 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2011 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Object representing a code comment.
+ * @author kchadha@scratch.mit.edu (Karishma Chadha)
+ */
+'use strict';
+
+goog.provide('Blockly.ScratchBlockComment');
+
+goog.require('Blockly.Comment');
+goog.require('Blockly.ScratchBubble');
+goog.require('Blockly.Events.BlockChange');
+goog.require('Blockly.Events.Ui');
+goog.require('Blockly.Icon');
+goog.require('goog.userAgent');
+goog.require('goog.math.Coordinate');
+
+
+/**
+ * Class for a comment.
+ * @param {!Blockly.Block} block The block associated with this comment.
+ * @param {?number} x Initial x position for comment
```

Comment:
Why are these all allowed to be undefined?

Reply:
These are allowed to be undefined because they are not strictly necessary for creating a comment.

A smart position will be chosen if x and y are not provided (based on the absolute position of the anchor on the block and the size of the stack that the block is part of). This is important for importing comments from 2.0 (during deserialization of scratch 2 projects, I'm relying on the auto layout by not passing comment positioning in the XML).